### ISSUE TITLE
Feauture Request: also provide models in torchscript format.

### ISSUE THREAD

**Issue by SR-Lut3t1um** [collabora/WhisperSpeech#104]
Request in title.

Would be really helpful to integrate WhisperSpeech easily into e.g. android.

**Comment by zoq** [comment]
Technically this should be possible, I marked this as a good first issue, in case someone likes to take a closer look into this.

**Comment by jpc** [comment]
It may be worth looking into the recent [ExecuTorch](https://pytorch.org/executorch/stable/intro-how-it-works.html) as a possible alternative.

WhisperSpeech already works with `torch.compile` but I think we had some issues with `torch.export` around our implementation of RoPE.

**Comment by BBC-Esq** [comment]
This also worked:

```ffmpeg -hide_banner -loglevel panic -i "test 1.mp3" -threads 1 -acodec pcm_s16le -ac 1 -af aresample=resampler=soxr -ar 16000 "test.wav" -y```

**Comment by SR-Lut3t1um** [comment]
It seems to be that Executorch isn't compatible with Androids NNAPI. However I'm unsure if that leads to worse performance.

**Comment by gongouveia** [comment]
Hello, @SR-Lut3t1um is exporting as .onnx a good alternative?

**Comment by SR-Lut3t1um** [comment]
Well, according to the [pytorch docs](https://pytorch.org/tutorials/recipes/model_preparation_android.html) torchscript is required, so I'm afraid. onnx isn't going to be an alternative for that use case.

**Comment by gongouveia** [comment]
I am pretty sure it is possible to use ONNX Runtime and the NNAPI execution provider.
perhaps converting to ONNX may be less of a hassle

**Comment by SR-Lut3t1um** [comment]
Yeah onnx should be fine too

**Comment by BBC-Esq** [comment]
Here's a draft response for the GitHub issue:

---

Thanks for the detailed discussion here! Let me add some technical context from the current codebase that's relevant for anyone looking to tackle this.

**Known Challenges for TorchScript/Export:**

1. **RoPE Implementation** (`modules.py`) - The `Rotary` class uses dynamic caching with runtime checks (`if not self.seq_len_cached or seq_len > self.seq_len_cached`) and `torch.einsum`, which can be problematic for tracing. This is likely where the `torch.export` issues originated.

2. **FlexEmbeddings** - Has conditional paths between training/eval modes with `merged_in`/`merged_out` buffers that get populated by `convert_for_eval()`. The `forward()` and `unembed()` methods branch based on these.

3. **Dynamic shapes** - The models handle variable-length sequences with padding, and there's conditional reshaping in `embed_stoks()` based on `stoks_len`.

4. **External dependencies** - The Vocoder wraps Vocos (`vocos.Vocos`), which would also need to be exportable.

**Regarding ONNX:**

The suggestion to use ONNX with NNAPI execution provider is reasonable and might indeed be less friction than TorchScript. The `torch.onnx.export` path is often more forgiving than `torch.jit.script` for these patterns, and ONNX Runtime's NNAPI EP is well-supported.

**Suggested approach for contributors:**

1. Start with just the S2A model (`SADelARTransformer`) in eval mode after `optimize()` is called
2. Refactor `Rotary` to pre-compute sin/cos tables for a fixed max length
3. Use `torch.jit.trace` on the generation path rather than `torch.jit.script`
